llvm/lib/Target/RISCV/RISCVTargetMachine.cpp

```diff
@@ -570,6 +570,7 @@ void RISCVPassConfig::addPreEmitPass() {
     addPass(createMachineCopyPropagationPass(true));
   if (TM->getOptLevel() >= CodeGenOptLevel::Default)
     addPass(createRISCVLateBranchOptPass());
+  addPass(&BranchFolderPassID);
```

Comment:
It should have run already? I'd expect a "late branch opt" pass to avoid introducing new constructs that need cleanup

Reply:
The late branch optimisation pass is actually making it easier for branch folding to clean up things that copy propagation and other prior passes have introduced.